src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java

```diff
@@ -32,48 +38,96 @@ public RoleAssignmentsService(RoleService roleService) {
+                var success = assignRolesToUser(results, username, specification.getRolesToAssign());
+                if (success) {
+                    removeRolesFromUserAtCaseload(results, caseload, username, specification.getRolesToRemove());
                 }
-            });
+            }
+            var infoMap = Map.of(
+                    "prisonId", caseload,
+                    "numUsersMatched", String.valueOf(usernamesForCaseload.size()),
+                    "numAssignRoleSucceeded", String.valueOf(results.getNumAssignRoleSucceeded()),
+                    "numAssignRoleFailed", String.valueOf(results.getNumAssignRoleFailed()),
+                    "numUnAssignRoleSucceeded", String.valueOf(results.getNumUnAssignRoleSucceeded()),
+                    "numUnAssignRoleIgnored", String.valueOf(results.getNumUnAssignRoleIgnored()),
+                    "numUnAssignRoleFailed", String.valueOf(results.getNumUnAssignRoleFailed()));
+            telemetryClient.trackEvent("UpdateRollAssignment", infoMap,null);
+            caseloadResults.put(caseload,results);
         }
+        return caseloadResults;
     }
 
-    private void removeRolesFromUserAtCaseload(String caseload, String username, List<String> rolesToRemove) {
-        rolesToRemove.forEach(roleCodeToRemove -> {
-            try {
-                roleService.removeRole(username, caseload, roleCodeToRemove);
+    private boolean assignRolesToUser(final RoleAssignmentStats results, final String username, final List<String> rolesToAssign) {
+        boolean hasFail = false;
+        for (val roleCodeToAssign : rolesToAssign) {
+            if (!hasFail) {
+                val status = assignRole(username, roleCodeToAssign);
+                results.addAssignResult(status);
+                if(status == FAIL){
+                    hasFail = true;
+                }
+            } else {
+                results.addAssignResult(FAIL);
+            }
+        }
+        return !hasFail;
+    }
 
-            } catch (HttpClientErrorException.NotFound notFoundException) {
-                log.info("Username {} does not have role {} at caseload {}", username, roleCodeToRemove, caseload);
+    private Status assignRole(final String username, final String roleCodeToAssign) {
+        try {
+            roleService.assignRoleToApiCaseload(username, roleCodeToAssign);
+            return SUCCESS;
+
+        } catch (Exception e) {
+            val message = String.format("Failure while assigning roles %1$s to user %2$s. No roles will be removed from this user. Continuing with next user.", roleCodeToAssign, username);
+            log.warn(message, e);
+            return FAIL;
+        }
+    }
+
+    private void removeRolesFromUserAtCaseload(final RoleAssignmentStats results, final String caseload, final String username, final List<String> rolesToRemove) {
+        boolean hasFail = false;
+        for (val roleCodeToRemove : rolesToRemove) {
```

Comment:
streams please

Reply:
Think we deleted the wrong comment.

This needs to be a foreach (unless we resort to an atomic boolean)... though thinking about it we could possibly look at using `reduce` instead.

The deleted comment was about flipping the boolean round we can have a look at tomorrow.

Reply:
maybe something like: 
```java
        return !rolesToAssign.stream().reduce(false, (hasFailed, role) -> {
            if (hasFailed) {
                results.addAssignResult(FAIL);
                return true;
            }
            val status = assignRole(username, role);
            results.addAssignResult(status);
            return status == FAIL;
        }, Boolean::logicalAnd);
```
(if we are happy with streaming and side effects)

Reply:
altered method to use stream reduce as suggested